src/core/src/core_logic/RebootManager.py

```diff
@@ -14,116 +14,167 @@
+        self.status_handler.set_installation_reboot_status(Constants.RebootStatus.STARTED)
+        reboot_init_time = self.env_layer.datetime.datetime_utcnow()
+
+        # Reboot after system-wide notification broadcast - no new logins will be allowed after this point.
+        self.env_layer.reboot_machine(self.__reboot_cmd + str(Constants.REBOOT_NOTIFY_TIMEOUT_IN_MINUTES) + ' ' + message)
+
+        # Safety net - if the machine doesn't reboot, we need to fail the operation.
+        max_allowable_time_to_reboot_in_minutes = self.__calc_max_allowable_time_to_reboot_in_minutes(maintenance_window_available_time_in_minutes)
+        while 1:
+            current_time = self.env_layer.datetime.datetime_utcnow()
+            elapsed_time_in_minutes = self.env_layer.datetime.total_minutes_from_time_delta(current_time - reboot_init_time)
+
+            # Keep logging to indicate machine hasn't rebooted yet. If successful, this will be the last log we see from this process.
+            if elapsed_time_in_minutes < max_allowable_time_to_reboot_in_minutes:
+                self.__reboot_wait_pulse(int(elapsed_time_in_minutes), int(max_allowable_time_to_reboot_in_minutes), int(maintenance_window_available_time_in_minutes))
+                continue
+
+            # If we get here, the machine has not rebooted in the time we expected. We need to fail the operation.
+            # This may be because of the following reasons:
+            # 1. The machine is not responding to the reboot command because of a customer environment issue. (customer should retry after a forcing a control-plane reboot)
+            # 2. The reboot command was externally interrupted during the broadcast period. (customer should retry after a forcing a control-plane reboot)
+            # 3. The time required to handle changes prior to reboot is greater than the time we've allocated. (action on AzGPS if seen at scale in Azure)
+            self.status_handler.set_installation_reboot_status(Constants.RebootStatus.FAILED)
+            error_msg = "Customer environment issue: Reboot failed to proceed on the machine in a timely manner. Please retry the operation."
+            self.status_handler.add_error_to_status(error_msg, Constants.PatchOperationErrorCodes.DEFAULT_ERROR)
+            raise Exception(error_msg, "[{0}]".format(Constants.ERROR_ADDED_TO_STATUS))
+
+    def __reboot_wait_pulse(self, elapsed_time_in_minutes, max_allowable_time_to_reboot_in_minutes, maintenance_window_available_time_in_minutes):
+        # type: (int, int, int) -> None
+        self.composite_logger.log("[RM] Waiting for machine reboot. [ElapsedTimeInMinutes={0}][MaxTimeInMinutes={1}][MaintenanceWindowAllowable={2}]"
+                                  .format(str(elapsed_time_in_minutes), str(max_allowable_time_to_reboot_in_minutes), str(maintenance_window_available_time_in_minutes - elapsed_time_in_minutes)))
```

Comment:
str(maintenance_window_available_time_in_minutes - elapsed_time_in_minutes)

This should come from the function call (i.e. line 145) for maintenance_window_available_time_in_minutes

Reply:
Addressed and renamed variable for clarity